### PR TITLE
[chainerx] Mean Var Python Bindings

### PR DESCRIPTION
**Pull request by kshitij12345**
Expose the present Mean and Var implementations to Python via Bindings.

Changes
* Return default float as the output for integral types.
* Update documentation.
* Tests (I think this test code has a lot of replication. Can someone please help with the best way to write these tests.)

### CODE REVIEW

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 1]
```diff
@@ -1292,3 +1292,145 @@ def test_max(is_module, xp, device, input, axis, dtype):
...
+    ])
+    @pytest.mark.parametrize('dtypes', _expected_dtypes_math_functions)
+    @pytest.mark.parametrize_device(['native:0', 'cuda:0'])
+    # TODO: Remove strides_check=False
+    def test_invalid_mean(self, xp, device, input, axis, dtypes):
+        return self.apply_func(xp, device, input, axis, dtypes)
+
+
+class TestVar(object):
+    def apply_func(self, xp, device, input, axis, dtypes):
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Does this have to be a method? You could make it a free function and reuse it for `mean` and `var` test. In that case, you don't need these "Test" prefixed classes neither.

**Comment by hvy** [comment]
Can you add the `ndarray.{mean,var}` bindings too? They are often tested using a `is_module` flag such as this one https://github.com/chainer/chainer/blob/fed13618562b05ccb18cf571ddb52c3714097358/tests/chainerx_tests/unit_tests/routines_tests/test_math.py#L722

File: chainerx_cc/chainerx/routines/statistics.cc [hunk 2]
```diff
@@ -7,14 +7,21 @@
...
 #include "chainerx/backward_builder.h"
 #include "chainerx/backward_context.h"
+#include "chainerx/dtype.h"
 #include "chainerx/macro.h"
 #include "chainerx/routines/creation.h"
+#include "chainerx/routines/type_util.h"
 
 namespace chainerx {
 
+Dtype promote_int2float(Dtype dtype){
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply chainerx_cc/chainerx/routines/statistics.cc]: Oh sure.

File: chainerx/_docs/routines.py [hunk 3]
```diff
@@ -1134,6 +1134,55 @@ def _docs_statistics():
...
+
+Returns:
+    :class:`~chainerx.ndarray`: The mean of ``a``, along the axis or axes if
+    specified.
+
+.. seealso:: :func:`numpy.mean`
+""")
+
+
+def _docs_statistics():
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply chainerx/_docs/routines.py]: Oops. Perks of copying.

File: chainerx_cc/chainerx/routines/statistics.cc [hunk 4]
```diff
@@ -48,7 +55,9 @@ Array Mean(const Array& a, const OptionalAxes& axis, bool keepdims) {
 Array Var(const Array& a, const OptionalAxes& axis, bool keepdims) {
     // TODO(hvy): Consider allowing device implementations.
     Axes sorted_axis = internal::GetSortedAxesOrAll(axis, a.ndim());
-    Array diff = a - Mean(a, sorted_axis, true);
+    Dtype mean_out = promote_int2float(a.dtype());
+    // TODO: remove once subtract allows mixed types.
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply chainerx_cc/chainerx/routines/statistics.cc]: Sure.

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 5]
```diff
@@ -1292,3 +1292,145 @@ def test_max(is_module, xp, device, input, axis, dtype):
...
+        in_dtype, out_dtype = dtypes
+        try:
+            a_np = input.astype(in_dtype)
+        except (ValueError, OverflowError):
+            return xp.zeros(())  # invalid combination of data and dtype
+
+        a = xp.array(a_np)
+        if xp is numpy:
+            a = dtype_utils.cast_if_numpy_array(xp, a, out_dtype)
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Yes. Have done that in the new commits.

File: chainerx/_docs/routines.py [hunk 6]
```diff
@@ -1134,6 +1134,55 @@ def _docs_statistics():
...
+        chainerx.mean,
+        """mean(a, axis=None, keepdims=False)
+Compute the arithmetic mean along the specified axis.
+
+Returns the average of the array elements. The average is taken over the flattened array by default,
+otherwise over the specified axis.
+
+Args:
+    a (~chainerx.ndarray): Array to take the mean of.
+    axis (None or int): Along which axis or axes to compute the mean. The flattened
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply chainerx/_docs/routines.py]: Done.

File: chainerx/_docs/routines.py [hunk 7]
```diff
@@ -1134,6 +1134,55 @@ def _docs_statistics():
...
+        chainerx.var,
+        """var(a, axis=None, keepdims=False)
+Compute the arithmetic var along the specified axis.
+
+Returns the var of the array elements. The var is taken over the flattened array by default,
+otherwise over the specified axis.
+
+Args:
+    a (~chainerx.ndarray): Array to take the var of.
+    axis (None or int): Along which axis or axes to compute the var. The flattened
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply chainerx/_docs/routines.py]: Done.

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py (hunk 1 above)
Comment (kshitij12345) [reply tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Have refactored the test code for it to be a function. Please a look

**Comment by kshitij12345** [comment]
Have expose `mean` and `var` as `ndarray` bindings and updated the tests for the same.

**Comment by hvy** [comment]
Thanks, it basically looks good to me. Let me trigger the CI.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Could you fix the style error? https://jenkins.preferred.jp/job/chainer/job/chainer_pr/86/TEST=CHAINERX_misc,sakura11=sakura11/console

**Comment by kshitij12345** [comment]
Have fixed it.

**Comment by hvy** [comment]
Jenkins, test this please.

File: chainerx_cc/chainerx/routines/statistics.cc [hunk 8]
```diff
@@ -7,14 +7,21 @@
...
+#include "chainerx/dtype.h"
 #include "chainerx/macro.h"
 #include "chainerx/routines/creation.h"
+#include "chainerx/routines/type_util.h"
 
 namespace chainerx {
 
+Dtype PromoteInt2Float(Dtype dtype){
+    return GetKind(dtype) == DtypeKind::kFloat? dtype :internal::GetDefaultDtype(DtypeKind::kFloat);
+}
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/statistics.cc]: Seems like this function should be clang-formatted (to a single line).

Comment (kshitij12345) [reply chainerx_cc/chainerx/routines/statistics.cc]: Updated.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Seems like `test_math.py` has been rewritten a bit. Could you rebase or merge with master? You can check the CI failures on Travis also, e.g. that `_expected_dtypes_math_functions` now is undefined.

**Comment by kshitij12345** [comment]
Oh right. On it.

**Comment by kshitij12345** [comment]
Have updated the tests. Please review.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Could you test backward and double backward in the Python tests? It's done with the other routines in `test_math.py` e.g. https://github.com/chainer/chainer/blob/master/tests/chainerx_tests/unit_tests/routines_tests/test_math.py#L1374.

**Comment by kshitij12345** [comment]
* Have rebased.
* Have added test to check the backward and double backward.
* However do note that the previous tests haven't been removed as they check axis behaviour for various shapes and input axis.

**Comment by hvy** [comment]
We just had a release and a following code freeze, sorry about the delay. I think the implementation looks good but again is conflicting. Could you fix it?

**Comment by kshitij12345** [comment]
Oh that is nice. Congrats.

On it.

**Comment by kshitij12345** [comment]
Errors seem non-relevant . Please Review

**Comment by hvy** [comment]
Sorry, again conflicting.

**Comment by kshitij12345** [comment]
@hvy Have rebased. Please review.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
CI failures seem irrelevant. Thanks for the long running PR and LGTM!

**Pull request merged**